### ISSUE TITLE
Leaked Semaphore objects

### ISSUE THREAD

**Issue by samLozier** [sqlfluff/sqlfluff#4080]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When running sqlfluff with multiprocessing, I frequently get leaked semaphore warnings when sqlfluff finishes. I haven't identified what causes this exactly, but I'm wondering if it happens when I kill a sqlfluff run and sqlfluff doesn't shut down all the processes it started (eg ctrl+c). 

### Expected Behaviour

No leakage of seamphore objects.

### Observed Behaviour

```/usr/local/Cellar/python@3.9/3.9.13_4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 104 leaked semaphore objects to clean up at shutdown
```

### How to reproduce

As I mentioned above, I'm not exactly sure what's been triggering this, but my best guess is that I get the warning after this sequence: 
1. start sqlfulff
2. kill sqlfluff with ctrl+c
3. start sqlfluff again
4. sqlfluff finishes successfully and produces this leaked semaphore warning

What i'm not sure of is if the warning is coming from the second successful run, or is a leftover error from the first run. 

### Dialect

Snowflake 

### Version

sqlfluff, version 1.4.1
sqlfluff-templater-dbt 1.4.1

### Configuration

[sqlfluff]
exclude_rules = L031, L033, L034, L023
select_clause_trailing_comma="forbid"
verbose = 0
nocolor = False
dialect = snowflake
templater = dbt
rules = None
recurse = 0
output_line_length = 120
runaway_limit = 10
ignore_templated_areas = True
encoding = autodetect
processes = 0
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:templater:dbt]
# project_dir = .
# profiles_dir = ./profiles/
# profile = snowflake-db

[sqlfluff:indentation]
indented_joins = True
indented_using_on = True
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = True

[sqlfluff:templater:jinja]
apply_dbt_builtins = True
load_macros_from_path= macros/

[sqlfluff:templater:jinja:macros]
# Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable, default='') %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}True{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 100
indent_unit = space
# comma_style = leading
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L003]
indent_unit=space

# [sqlfluff:rules:L007]  # Keywords
# operator_new_lines = after

[sqlfluff:layout:type:binary_operator]  # Keywords
line_position = leading

[sqlfluff:layout:type:comma]  # Keywords
line_position = leading

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = consistent

[sqlfluff:rules:L011]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L012]  # Aliasing
aliasing = explicit

[sqlfluff:rules:L014]  # Unquoted identifiers
unquoted_identifiers_policy = all
extended_capitalisation_policy = upper

[sqlfluff:rules:L016]
ignore_comment_lines = True
maximum_line_length = 120

[sqlfluff:rules:L026]
force_enable = False

[sqlfluff:rules:L028]
force_enable = False

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = aliases
quoted_identifiers_policy = none

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = upper

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = consistent

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = join

[sqlfluff:rules:L044]

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_1 = False
prefer_count_0 = False

[sqlfluff:rules:L052]  # Semi-colon formatting approach.
semicolon_newline = False
require_final_semicolon = False

[sqlfluff:rules:L054]  # GROUP BY/ORDER BY column references.
group_by_and_order_by_style = explicit

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by greg-meinecke** [comment]
Hitting a similar issue using sqlfluff with pre-commit. Configs below:

`.pre-commit-config.yaml`
```
- repo: https://github.com/sqlfluff/sqlfluff
  rev: 2.0.2
  hooks:
    - id: sqlfluff-lint
    - id: sqlfluff-fix
```

`.sqlfluff`
```
[sqlfluff]
sql_file_exts = .sql
max_line_length = 120
dialect = postgres

[sqlfluff:indentation]
indented_joins = false
indented_using_on = true
template_blocks_indent = false

[sqlfluff:rules:capitalisation.keywords]
capitalisation_policy = upper
```

```
❯ pre-commit run --all-files
check yaml...............................................................Passed
check for added large files..............................................Passed
fix end of files.........................................................Passed
fix requirements.txt.....................................................Passed
trim trailing whitespace.................................................Passed
black (python)...........................................................Passed
isort (python)...........................................................Passed
flake8 (python)..........................................................Passed
Terraform fmt............................................................Passed
sqlfluff-lint............................................................Failed
- hook id: sqlfluff-lint
- exit code: 1

== [somefile.sql] FAIL                              
L: 122 | P:   1 | LT05 | Line is too long (156 > 120).
                       | [layout.long_lines]
L: 126 | P:   5 | LT05 | Line is too long (133 > 120).
                       | [layout.long_lines]
L: 149 | P:   5 | LT05 | Line is too long (131 > 120).
                       | [layout.long_lines]
L: 155 | P:   5 | LT05 | Line is too long (139 > 120).
                       | [layout.long_lines]
L: 161 | P:   5 | LT05 | Line is too long (139 > 120).
                       | [layout.long_lines]
L: 193 | P:   1 | LT05 | Line is too long (122 > 120).
                       | [layout.long_lines]
L: 194 | P:   1 | LT05 | Line is too long (127 > 120).
                       | [layout.long_lines]
L: 212 | P:   1 | LT05 | Line is too long (135 > 120).
                       | [layout.long_lines]
L: 220 | P:   1 | LT05 | Line is too long (141 > 120).
                       | [layout.long_lines]
All Finished 📜 🎉!                                                                
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/mp-m6yyuylp': [Errno 2] No such file or directory
  warnings.warn('resource_tracker: %r: %s' % (name, e))

sqlfluff-fix.............................................................Failed
- hook id: sqlfluff-fix
- exit code: 1

==== finding fixable violations ====
==== no fixable linting violations found ====                                      
All Finished 📜 🎉!
  [9 unfixable linting violations found]
==== lint for unfixable violations ====
== [somefile.sql] PASS
== [somefile.sql] FAIL
L: 122 | P:   1 | LT05 | Line is too long (156 > 120). [layout.long_lines]
L: 126 | P:   5 | LT05 | Line is too long (133 > 120). [layout.long_lines]
L: 149 | P:   5 | LT05 | Line is too long (131 > 120). [layout.long_lines]
L: 155 | P:   5 | LT05 | Line is too long (139 > 120). [layout.long_lines]
L: 161 | P:   5 | LT05 | Line is too long (139 > 120). [layout.long_lines]
L: 193 | P:   1 | LT05 | Line is too long (122 > 120). [layout.long_lines]
L: 194 | P:   1 | LT05 | Line is too long (127 > 120). [layout.long_lines]
L: 212 | P:   1 | LT05 | Line is too long (135 > 120). [layout.long_lines]
L: 220 | P:   1 | LT05 | Line is too long (141 > 120). [layout.long_lines]
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
/opt/homebrew/Cellar/python@3.10/3.10.9/Frameworks/Python.framework/Versions/3.10/lib/python3.10/multiprocessing/resource_tracker.py:237: UserWarning: resource_tracker: '/mp-ju0qrgwj': [Errno 2] No such file or directory
  warnings.warn('resource_tracker: %r: %s' % (name, e))
```

Please let me know if I can provide any more information to help. Thanks!

**Comment by alanmcruickshank** [comment]
Looks like this is now resolved upstream in cpython: https://github.com/python/cpython/issues/90549

Based on the python release schedule, I would expect (hope?) that this fix made it into python 3.11 (if not then almost certainly python 3.12). For anyone who's getting this bug, please confirm whether upgrading to python 3.11 and/or 3.12 resolves it?

**Comment by peterldowns** [comment]
Still seeing this warning, using sqlfluff version 3.0.5 installed via nix, which uses python 3.11 under the hood

https://github.com/NixOS/nixpkgs/blob/74948618db12dac53fb3392ecf423fa948b72b03/pkgs/development/tools/database/sqlfluff/default.nix#L8

```console
$ which sqlfluff
/nix/store/y6fslcdicj4svdqyw97my7lwd762g6xp-sqlfluff-3.0.5/bin/sqlfluff

$ tail /nix/store/y6fslcdicj4svdqyw97my7lwd762g6xp-sqlfluff-3.0.5/bin/sqlfluff
PATH=${PATH%':'}
export PATH
PATH=${PATH:+':'$PATH':'}
PATH=${PATH/':''/nix/store/mqdc5233rdz150n1crdyl2cirpa7b6qg-python3-3.11.9/bin'':'/':'}
PATH='/nix/store/mqdc5233rdz150n1crdyl2cirpa7b6qg-python3-3.11.9/bin'$PATH
PATH=${PATH#':'}
PATH=${PATH%':'}
export PATH
export PYTHONNOUSERSITE='true'
exec -a "$0" "/nix/store/y6fslcdicj4svdqyw97my7lwd762g6xp-sqlfluff-3.0.5/bin/.sqlfluff-wrapped"  "$@"
```

**Comment by peterldowns** [comment]
Even if upstream is what's emitting these warnings, I believe sqlfluff could use the `warnings` package to prevent these errors from being shown. If desired, it could be that they are shown by default unless a `--no-python-warnings` flag is passed (or something like that.)

https://docs.python.org/3/library/warnings.html

For anyone else who'd like to work around this, you can set `PYTHONWARNINGS=ignore` to hide the errors.

**Comment by ramtinJ95** [comment]
This popped up for me today, im on 3.4.2 using python 3.13.7.When i do sqlfluff lint -p and some number larger than 1. Not on apple chips either!

**Comment by fizyk** [comment]
Same thing happened to me, sqlfluff 4.0.4, python 3.14.3

```
/usr/local/lib/python3.14/multiprocessing/resource_tracker.py:396: UserWarning: resource_tracker: There appear to be 5 leaked semaphore objects to clean up at shutdown: {'/mp-9rbpcpcn', '/mp-d7a8mf_t', '/mp-9kgf0vci', '/mp-iaujcm4e', '/mp-7q6b5j6f'}
  warnings.warn(
/usr/local/lib/python3.14/multiprocessing/resource_tracker.py:412: UserWarning: resource_tracker: '/mp-d7a8mf_t': [Errno 2] No such file or directory
  warnings.warn('resource_tracker: %r: %s' % (name, e))
/usr/local/lib/python3.14/multiprocessing/resource_tracker.py:412: UserWarning: resource_tracker: '/mp-7q6b5j6f': [Errno 2] No such file or directory
  warnings.warn('resource_tracker: %r: %s' % (name, e))
```

**Comment by keraion** [comment]
Unfortunately, I'm still seeing this issue just running in parallel over the fixture dialects
```sh
sqlfluff lint test/fixtures/dialects -p8

...

/usr/lib/python3.13/multiprocessing/resource_tracker.py:276: UserWarning: resource_tracker: There appear to be 3 leaked semaphore objects to clean up at shutdown: {'/mp-731i61l1', '/mp-t62vmkfh', '/mp-uybdsylp'}
  warnings.warn(
```
